# 0. Initial Set-up

In [1]:
import numpy as np
import pandas as pd
import requests
import os
import json
import tqdm

from rdkit import Chem
from tqdm import tqdm
from thermo import functional_groups
from Bio import Entrez
from chembl_structure_pipeline import checker

from rdkit.Chem import rdMolDescriptors, Descriptors, Lipinski, Crippen, inchi
from rdkit.Chem.FilterCatalog import FilterCatalog, FilterCatalogParams
from concurrent.futures import ThreadPoolExecutor, as_completed

import utils
import filters
from data_gathering import download_and_save

[13:18:09] Initializing Normalizer


In [2]:
# Save the path to your curation folder to "curation_path"
dataset_name = 'AID1798' 
curation_path = 'S:/coding/WelQrate/' 
data_folder = f'{curation_path}/data/{dataset_name}'

# Columns to be extracted from the assays:
# Modify if your datasets have different format
smi_col = 'PUBCHEM_EXT_DATASOURCE_SMILES' # Column containing SMILES
cid_col = 'PUBCHEM_CID' # Column containing identifiers (e.g, CIDs)
activity_col = 'PUBCHEM_ACTIVITY_OUTCOME' # Column containing activity outcomes
col_list = [cid_col, smi_col, activity_col]

# 1. Data Gathering

Before importing data, need to identify which AIDs will be included. 

Data will be imported from https://pubchem.ncbi.nlm.nih.gov/assay/. For more information on PubChem's programmatic access, refer to: https://pubchem.ncbi.nlm.nih.gov/docs/bioassays. Some other programmatic access options available such as PUG-REST. However, these might not be optimal for bulk retrieval or handling of large dataset due to the limitation of request volume.

Data for individual assays include 7 required columns (CIDs, isomeric SMILES, etc.) and optional test results. Refer to https://ftp.ncbi.nlm.nih.gov/pubchem/Bioassay/CSV/README for further details. For datasets intended for regression model, additional columns could be extracted accordingly.

In [ ]:
# Desired AIDs
AIDs = [626, 1488, 1741]

In [ ]:
#Keep unique values in list AIDs (since there could be overlapping AIDs from different targets or project)
AIDs = list(set(AIDs))
AIDs = [str(AID) for AID in AIDs]
print('Number of datasets retrieving: ', len(AIDs))

Number of datasets retrieving:  3


In [21]:
download_and_save(AIDs, data_folder, col_list, smi_col, cid_col, activity_col)

Completed 1 out of 3 datasets.
Completed 2 out of 3 datasets.
Completed 3 out of 3 datasets.


# 2. Isomeric SMILES

For the purpose of our project, we would like to include isomeric form of SMILES representation in our final dataset. Although PubChem claimed that their datatable should include isomeric SMILES (https://pubchem.ncbi.nlm.nih.gov/docs/bioassays), some dataset might include non-isomeric SMILES. This step is to import isomeric SMILES based on CIDs.

Several packages such as RDkit have modules to return isomeric SMILES from a given input SMILES. However, for consistency, we decided to use the PubChem Identifier Exchange Service, which take an input identifier (CIDs, SMILES, InChI, etc.)  and return the corresponding identifier (CIDs, isomeric SMILES, InChIs, etc.). Here, we export the list of CIDs for compounds in our dataset and use this server to retrieve their isomeric SMILES. For more information, refer to: https://pubchem.ncbi.nlm.nih.gov/docs/identifier-exchange-service

In [ ]:
#Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_2'):
    os.makedirs(f'{data_folder}/before_finished/step_2')

#Export list of CIDs to csv with one column (without the column name):
for AID in AIDs:
    assay = pd.read_csv(f'{data_folder}/before_finished/step_1/AID{AID}.csv')
    cids = assay['PUBCHEM_CID'].astype(int)  # Ensure the CIDs are integers
    cids.to_csv(f'{data_folder}/before_finished/step_2/CID{AID}.csv', index=False, header=False)

#After this step, we submit the list at https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi with operator type "same CID" and Output IDs "SMILES" (isomeric SMILES by default)
#https://pubchem.ncbi.nlm.nih.gov/docs/identifier-exchange-service for more details
#Here I named the output file as "SMILES{AID}.txt"

In [ ]:
def check_isomeric_smiles(AIDs):
    """
    Check if the SMILES in the assay are the same as the isomeric forms returned by pubchem idexchange.
    Input: AIDs (list of strings)
    Output: non_isomeric_smi_cids (dictionary with AID as key and list of CIDs as values for the datasets in AIDs
    """
    non_isomeric_smi_cids = {}
    for AID in AIDs:    
        non_isomeric_smi_cids[AID] = []
        #import SMILES.txt file as a table:
        correct_isomeric_smiles = pd.read_csv(f'{data_folder}/before_finished/step_2/isomeric_smi_{AID}.txt', sep='\t', header=None)
        assay = pd.read_csv(f'{data_folder}/before_finished/step_1/AID{AID}.csv')

        #compare smiles in assay with smiles in correct_smiles:
        for cid in assay['PUBCHEM_CID']:
            if assay.loc[assay['PUBCHEM_CID'] == cid, 'PUBCHEM_EXT_DATASOURCE_SMILES'].values[0] != correct_isomeric_smiles.loc[correct_isomeric_smiles[0] == cid, 1].values[0]:
                non_isomeric_smi_cids[AID].append(cid)

        if len(non_isomeric_smi_cids[AID]) == 0:
            print(f'All SMILES in AID {AID} are isomeric')
        else:
            print(f'There are some potential non-isomeric SMILES in AID {AID}:')
            print(non_isomeric_smi_cids[AID])

    return non_isomeric_smi_cids

def update_isomeric(AIDs, non_isomeric_smi_cids):
    """
    Update the SMILES in the assay to isomeric SMILES.
    Input: AIDs (list of strings), non_isomeric_smi_cids (dictionary with AID as key and list of non-isomeric CIDs as values)
    """
    with open(f'{data_folder}/before_finished/step_2/non_isomeric_smi_cids.txt', 'w') as f:
        # record the non-isomeric SMILES 
        for AID in AIDs:
            assay = pd.read_csv(f'{data_folder}/before_finished/step_1/AID{AID}.csv')
            correct_isomeric_smiles = pd.read_csv(f'{data_folder}/before_finished/step_2/isomeric_smi_{AID}.txt', sep='\t', header=None)
            f.write(f'AID {AID}: {non_isomeric_smi_cids[AID]}\n')

            for cid in non_isomeric_smi_cids[AID]:
                f.write(f'CID {cid}: {assay.loc[assay["PUBCHEM_CID"] == cid, "PUBCHEM_EXT_DATASOURCE_SMILES"].values[0]} -> {correct_isomeric_smiles.loc[correct_isomeric_smiles[0] == cid, 1].values[0]}\n')
                assay.loc[assay['PUBCHEM_CID'] == cid, 'PUBCHEM_EXT_DATASOURCE_SMILES'] = correct_isomeric_smiles.loc[correct_isomeric_smiles[0] == cid, 1].values[0]

            f.write(f'===\n')
            assay.to_csv(f'{data_folder}/before_finished/step_2/AID{AID}.csv', index=False)       

In [ ]:
non_isomeric_smi_cids = check_isomeric_smiles(AIDs)

All SMILES in AID 1488 are isomeric
There are some non-isomeric SMILES in AID 626:
[2997662, 2997957, 2999888]
All SMILES in AID 1741 are isomeric


Note: Here they returned that three smiles in AID626 were not isomeric. Again, this shows that the SMILES representation of some compounds in the given datasets might not be isomeric.

In [ ]:
update_isomeric(AIDs, non_isomeric_smi_cids)

# 3. Import InChIs

We would like to include standard InChI to diversify users' choice of which data they would like to use for their own benchmark.

In [ ]:
#Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_3'):
    os.makedirs(f'{data_folder}/before_finished/step_3')

Again, it is convenient to use the PubChem Identifier Exchange Service (https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi) with operator type "same CID" and Output IDs "InChI" to retrieve InChI from a given list of input CIDs. The same CID lists from STEP 2 could be used here. The resulted InChIs could be checked if being standard by indentifying the presence of 'InChI=1S' at the begining of each InChI string.

In [ ]:
"""
CID lists (in "step_2" folder should be submitted to PubChem Identifier Exchange Service)
    Operator type: "same CID" 
    Output IDs "InChI"
    Output method: "Two column file showing each input output-correspondence"
    Compression: "No compression"
InChI list should be saved into "step_3" folder, named as "std_inchi_{AID}.txt" 
"""
# Import dataframes:
for AID in AIDs: 
    exec(f'AID{AID} = pd.read_csv("{data_folder}/before_finished/step_2/AID{AID}.csv")')
    exec(f'AID{AID}_InChI = pd.read_csv("{data_folder}/before_finished/step_3/std_inchi_{AID}.txt", sep="\\t", header=None)')

In [ ]:
#Check if they are all standard InChI:
for AID in AIDs:
    check_inchi = f"""
non_standard_InChI = []
for i in range(len(AID{AID}_InChI[1])):
    if not AID{AID}_InChI[1][i].startswith('InChI=1S'):
        non_standard_InChI.append(AID{AID}_InChI[1][i])
if not non_standard_InChI:
    print('All InChI in AID{AID} are standard')
else:
    print('There are some non-standard InChI in AID{AID}')
    print(non_standard_InChI)
    print('===')
"""
    exec(check_inchi)

All InChI are standard


Now we concatenate the InChIs in our tables:

In [ ]:
# Update and save the files
for AID in AIDs: 
    update_inchi = f"""
AID{AID}_InChI_dict = dict(zip(AID{AID}_InChI[0], AID{AID}_InChI[1]))
AID{AID}['InChI'] = AID{AID}[cid_col].map(AID{AID}_InChI_dict)
AID{AID}['InChI'] = AID{AID}['InChI'].astype(str)
AID{AID}[cid_col] = AID{AID}[cid_col].astype(int)
AID{AID}.to_csv(r"{data_folder}/before_finished/step_3/AID{AID}.csv", index=False)
"""
    exec(update_inchi)

# 4. Check Duplicates

When checking duplicates in the datasets, we would like to know if there are
1) Multiple identical molecules
2) Molecules with identical CID but different InChIs or SMILES
3) Molecules with identical InChI but with different CIDs or SMILES

In [ ]:
#import:
for AID in AIDs:
    exec(f"AID{AID} = pd.read_csv(r'{data_folder}/before_finished/step_3/AID{AID}.csv', sep=',', header=0)")

#Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_4'):
    os.makedirs(f'{data_folder}/before_finished/step_4')

## 4.1. Checking identical molecules

In [ ]:
#Return all duplicates by comparing InChI, SMILES, and CIDs:
for AID in AIDs:
    check_duplicate = f"""
AID{AID}_duplicates_InChI = AID{AID}[AID{AID}.duplicated(subset=['InChI'], keep=False)]
AID{AID}_duplicates_SMILES = AID{AID}[AID{AID}.duplicated(subset=[smi_col], keep=False)]
AID{AID}_duplicates_CIDs = AID{AID}[AID{AID}.duplicated(subset=[cid_col], keep=False)]
print('Number of AID{AID} InChI duplicates: ', len(AID{AID}_duplicates_InChI))
print('Number of AID{AID} SMILES duplicates: ', len(AID{AID}_duplicates_SMILES))
print('Number of AID{AID} CID duplicates: ', len(AID{AID}_duplicates_CIDs))
"""
    exec(check_duplicate)

In [ ]:
#write duplicates to a txt file: 
with open(f'{data_folder}/before_finished/step_4/duplicates.txt', 'w') as f:
    for AID in AIDs: 
        duplicates_InChI = eval(f'AID{AID}_duplicates_InChI')
        duplicates_SMILES = eval(f'AID{AID}_duplicates_SMILES')
        duplicates_CIDs = eval(f'AID{AID}_duplicates_CIDs')
        f.write(f'\n\nAID{AID} InChI duplicates:\n')
        f.write(duplicates_InChI.to_string())
        f.write(f'\nAID{AID} SMILES duplicates:\n')
        f.write(duplicates_SMILES.to_string())
        f.write(f'\nAID{AID} CID duplicates:\n')
        f.write(duplicates_CIDs.to_string())

## 4.2. Same CIDs but different chemical representations

In [ ]:
#reindex
for AID in AIDs: 
    exec(f"AID{AID}_duplicates_CIDs.reset_index(drop=True, inplace=True)")

In [ ]:
with open(f'{data_folder}/before_finished/step_4/sameCID_different_others.txt', 'w') as f:
    for AID in AIDs: 
        sameCID_differentInChI = []
        sameCID_differentSMILES = []
        duplicates_CIDs = eval(f'AID{AID}_duplicates_CIDs')
        for i in range(len(duplicates_CIDs[cid_col])):
            for j in range(i+1, len(duplicates_CIDs[cid_col])):
                if duplicates_CIDs[cid_col][i] == duplicates_CIDs[cid_col][j]:
                    if duplicates_CIDs['InChI'][i] != duplicates_CIDs['InChI'][j]:
                        sameCID_differentInChI.append((duplicates_CIDs[cid_col][i], duplicates_CIDs[cid_col][j]))
                    if duplicates_CIDs[smi_col][i] != duplicates_CIDs[smi_col][j]:
                        sameCID_differentSMILES.append((duplicates_CIDs[cid_col][i], duplicates_CIDs[cid_col][j]))

        if sameCID_differentInChI == []:
            f.write(f'No duplicate CIDs with different InChIs in AID{AID}\n')
        else:
            f.write('Found duplicate CIDs with different InChIs in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameCID_differentInChI))
            f.write("\n")
        
        if sameCID_differentSMILES == []:
            f.write(f'No duplicate CIDs with different SMILES in AID{AID}\n')
        else:
            f.write(f'Found duplicate CIDs with different SMILES in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameCID_differentSMILES))
            f.write("\n")
        f.write("===\n")

No duplicate CIDs with different InChIs


## 4.3. Same InChI but with different CIDs or SMILES

In [ ]:
#reindex
for AID in AIDs: 
    exec(f"AID{AID}_duplicates_InChI.reset_index(drop=True, inplace=True)")

In [ ]:
with open(f'{data_folder}/before_finished/step_4/sameInChI_different_others.txt', 'w') as f:
    for AID in AIDs: 
        sameInChI_differentCID = []
        sameInChI_differentSMILES = []
        duplicates_InChI = eval(f'AID{AID}_duplicates_InChI')
        for i in range(len(duplicates_InChI['InChI'])):
            for j in range(i+1, len(duplicates_InChI['InChI'])):
                if duplicates_InChI['InChI'][i] == duplicates_InChI['InChI'][j]:
                    if duplicates_InChI[cid_col][i] != duplicates_InChI[cid_col][j]:
                        sameInChI_differentCID.append((duplicates_InChI[cid_col][i], duplicates_InChI[cid_col][j]))
                    if duplicates_InChI[smi_col][i] != duplicates_InChI[smi_col][j]:
                        sameInChI_differentSMILES.append((duplicates_InChI[cid_col][i], duplicates_InChI[cid_col][j]))
        
        if sameInChI_differentCID == []:
            f.write(f'No duplicate InChIs with different CIDs in AID{AID}\n')
        else:
            f.write('Found duplicate InChIs with different CIDs in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameInChI_differentCID))
            f.write("\n")
        
        if sameInChI_differentSMILES == []:
            f.write(f'No duplicate InChIs with different SMILES in AID{AID}\n')
        else:
            f.write(f'Found duplicate InChIs with different SMILES in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameInChI_differentSMILES))
            f.write("\n")
        f.write("===\n")

## 4.4. Same SMILES but with different CIDs or InChIs

In [ ]:
#reindex
for AID in AIDs: 
    exec(f"AID{AID}_duplicates_SMILES.reset_index(drop=True, inplace=True)")

In [ ]:
with open(f'{data_folder}/before_finished/step_4/sameSMILES_different_others.txt', 'w') as f:
    for AID in AIDs: 
        sameSMILES_differentCID = []
        sameSMILES_differentInChI = []
        duplicates_SMILES = eval(f'AID{AID}_duplicates_SMILES')
        for i in range(len(duplicates_SMILES[smi_col])):
            for j in range(i+1, len(duplicates_SMILES[smi_col])):
                if duplicates_SMILES[smi_col][i] == duplicates_SMILES[smi_col][j]:
                    if duplicates_SMILES[cid_col][i] != duplicates_SMILES[cid_col][j]:
                        sameSMILES_differentCID.append((duplicates_SMILES[cid_col][i], duplicates_SMILES[cid_col][j]))
                    if duplicates_SMILES['InChI'][i] != duplicates_SMILES['InChI'][j]:
                        sameSMILES_differentInChI.append((duplicates_SMILES[cid_col][i], duplicates_SMILES[cid_col][j]))
        
        if sameSMILES_differentCID == []:
            f.write(f'No duplicate SMILES with different CIDs in AID{AID}\n')
        else:
            f.write(f'Found duplicate SMILES with different CIDs in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameSMILES_differentCID))
            f.write("\n")
        
        if sameSMILES_differentInChI == []:
            f.write(f'No duplicate SMILES with different InChIs in AID{AID}\n')
        else:
            f.write(f'Found duplicate SMILES with different InChIs in AID{AID}:\n')
            f.write('\n'.join(str(item) for item in sameSMILES_differentInChI))
            f.write("\n")
        f.write("===\n")

## 4.4 Drop duplicates

When dropping duplicates, we will keep the first molecule in a pair or a group of duplicates. For example, if there are 12 duplicates (6 pairs), we keep 6 componds.

In [ ]:
# Keep only the first duplicate in the dataframes:
for AID in AIDs: 
    exec(f"AID{AID}.drop_duplicates(subset=['InChI'], keep='first', inplace=True)")

    last_check = f"""
if len(AID{AID}[AID{AID}.duplicated(subset=['InChI'], keep=False)]) == 0:
    print('No more duplicate InChI in AID{AID}')
else:
    print('There are still duplicate InChI in AID{AID}')   
    """
    exec(last_check)

In [ ]:
for AID in AIDs: 
    last_check = f"""
if len(AID{AID}[AID{AID}.duplicated(subset=[smi_col], keep=False)]) == 0:
    print('No more duplicate SMILES in AID{AID}')
else:
    print('There are still duplicate SMILES in AID{AID}')   
    """
    exec(last_check)

In [ ]:
for AID in AIDs: 
    last_check = f"""
if len(AID{AID}[AID{AID}.duplicated(subset=[cid_col], keep=False)]) == 0:
    print('No more duplicate CID in AID{AID}')
else:
    print('There are still duplicate CID in AID{AID}')   
    """
    exec(last_check)

In [ ]:
# Save the dataframes to csv:
for AID in AIDs: 
    exec(f"AID{AID}.to_csv(r'{data_folder}/before_finished/step_4/AID{AID}.csv', index=False)")

# 5. Hierarchical Curation

For the hierarchical curation, there are some rules:

(1) All assays used should be on the same or close species/cell lines. Optimally, they should also be from the same project/laboratory.

(2) Primary actives (PrA) will have a large false-positive rate. Therefore, they should be tested in follow-up confirmatory screens (optimally dose-reponse).

(3) Actives could be promiscuous. Therefore, it is optimal to have counter-screens on different targets to test specificity.

(4) For some projects, compounds were tested in multiple rounds. Therefore, assays often have hierarchical relations. From a single primary screen (Pr), active compounds (Pr_A) could be tested in multiple rounds of confirmatory screens (Cf_1, Cf_2, ..., Cf_final) or counter screens (Ct_1, Ct_2, etc.). Actives from confirmatory screens (Cf_actives) have a higher possibility of being true active. If an active compound is tested active in counter screens (Cf_actives), it is likely to be a promiscuous compound and should not be included.

(4) It is important to know the relationship between assays. Active sets from downstream screens always have a lower false-positive rate than active sets from upstream screens due to better assay technologies on a smaller set of compounds. Therefore, final hits should be taken from the intersection of the very last confirmatory assays, without tested active in any counter-screen:
Final hits = [Cf_final1_actives ∩ Cf_final2_actives ∩ ...] \ [Ct_1_actives ∪ Ct_2_actives ∪ ...]

However, if the confirmatory assays are unrelated (tested on different set of compounds), then we might have to take the union of their active sets instead of the intersections as in this formula.

(5) The hierarchical relations should be inspected carefully to see if follow-up confirmatory screens include extra compounds (Ex) that were not tested in earlier screens or tested inactive in earlier screens. If exist, these compounds require manual inspection.

(6) Final inactives should be taken from primary inactives (Pr_inactives) (not inconclusive, unspecified, or probes), plus extra compounds that were tested inactive in conformatory screens (Ex_inactives), if justified.
Final inactives = Pr_inactives ∪ Ex_inactives

## 5.1. Classify groups of compounds in each assay by activities

In [ ]:
path = f'{data_folder}/before_finished/step_4'
keynumbers = [626, 1488, 1741] # specify the keynumbers you want to import

for keynumber in keynumbers:
    filename = os.path.join(path, f'AID{keynumber}.csv')
    if os.path.exists(filename):
        df = pd.read_csv(filename, index_col=None, header=0)
        exec(f'AID{keynumber} = df')
        exec(f'AID{keynumber}_active = df[df[activity_col]=="Active"]')
        exec(f'AID{keynumber}_inactive = df[df[activity_col]=="Inactive"]')
        exec(f'AID{keynumber}_inconclusive = df[df[activity_col]=="Inconclusive"]')
        exec(f'AID{keynumber}_unspecified = df[df[activity_col]=="Unspecified"]')
        exec(f'AID{keynumber}_probe = df[df[activity_col]=="Probe"]')

In [ ]:
#Create a df with first column the variables name, and the second column the number of rows:
df = pd.DataFrame(columns=['AID', 'Tested Compounds', 'Active', 'Inactive', 'Inconclusive', 'Unspecified', 'Probe'])
for keynumber in keynumbers:
    exec(f'df.loc[len(df)] = ["AID{keynumber}", len(AID{keynumber}), len(AID{keynumber}_active), len(AID{keynumber}_inactive), len(AID{keynumber}_inconclusive), len(AID{keynumber}_unspecified), len(AID{keynumber}_probe)]')
df

AID  Tested Compounds  Active  Inactive  Inconclusive  Unspecified  \
0   AID626             63676    1938     61738             0            0   
1  AID1488              1665     309       940           416            0   
2  AID1741              1665     150      1195           320            0   

   Probe  
0      0  
1      0  
2      0

## 5.2. Check the hierachical relations

In [ ]:
# Check if compounds tested in coonformatory AID1488 was active in the primary screen 626
AID1488['PUBCHEM_CID'].isin(AID626_active['PUBCHEM_CID']).value_counts()

PUBCHEM_CID
True    1665
Name: count, dtype: int64

In [ ]:
# Check if compounds tested in coonformatory AID1488 was active in the primary screen 626
AID1741['PUBCHEM_CID'].isin(AID626_active['PUBCHEM_CID']).value_counts()

PUBCHEM_CID
True    1665
Name: count, dtype: int64

In [ ]:
# Check how many active compounds in AID1488 were tested active in AID1741.
AID1488_active['PUBCHEM_CID'].isin(AID1741_active['PUBCHEM_CID']).value_counts()

PUBCHEM_CID
False    188
True     121
Name: count, dtype: int64

In [ ]:
len(AID1741_active) - AID1488_active['PUBCHEM_CID'].isin(AID1741_active['PUBCHEM_CID']).value_counts()

PUBCHEM_CID
False   -38
True     29
Name: count, dtype: int64

There is no problem with this assay data. So now we take actives in AID 1488 that were inactive in AID1741 as potential hits, which renders 188 compounds. And we take all inactives in AID626 as potential inactives.

## 5.3. Export the data

In [ ]:
#Subtract the CIDs from AID1741_active from AID1488_active:
potential_hits_cids = AID1488_active[~AID1488_active['PUBCHEM_CID'].isin(AID1741_active['PUBCHEM_CID'])]

#Get all the rows from AID626 that are in final_hits:
potential_hits_M1_agonist = AID626[AID626['PUBCHEM_CID'].isin(potential_hits_cids['PUBCHEM_CID'])]

In [ ]:
#Get all the rows from AID626 that are inactive
potential_inactives_M1_agonist = AID626_inactive

In [ ]:
if not os.path.exists(f'{data_folder}/before_finished/step_5'):
    os.makedirs(f'{data_folder}/before_finished/step_5')

In [ ]:
#export the potential hits and inactives to csv:
potential_hits_M1_agonist.to_csv(f'{data_folder}/before_finished/step_5/potential_actives.csv', index=False)
potential_inactives_M1_agonist.to_csv(f'{data_folder}/before_finished/step_5/potential_inactives.csv', index=False)

# 6. RDkit Parse Check

In [ ]:
potential_actives = pd.read_csv(f'{data_folder}/before_finished/step_5/potential_actives.csv', sep=',', header=0)
potential_inactives = pd.read_csv(f'{data_folder}/before_finished/step_5/potential_inactives.csv', sep=',', header=0)

In [ ]:
problems_actives, cannot_parse_actives = filters.rdkit_parse(potential_actives, smi_col, cid_col)
problems_inactives, cannot_parse_inactives = filters.rdkit_parse(potential_inactives, smi_col, cid_col)

No problems detected
No problems detected


In [ ]:
#Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_6'):
    os.makedirs(f'{data_folder}/before_finished/step_6')

with open(f'{data_folder}/before_finished/step_6/problem_list_actives.txt', 'w') as f:
    f.write("Problems:\n")
    for item in problems_actives:
        f.write("%s\n" % item)
    f.write("Cannot parse:\n")
    for item in cannot_parse_actives:
        f.write("%s\n" % item)

with open(f'{data_folder}/before_finished/step_6/problem_list_inactives.txt', 'w') as f:
    f.write("Problems:\n")
    for item in problems_inactives:
        f.write("%s\n" % item)
    f.write("Cannot parse:\n")
    for item in cannot_parse_inactives:
        f.write("%s\n" % item)

Our dataset returned no problem or non-parsable molecule.

# 7. Inorganics Filter

In [ ]:
# Import data
potential_actives = pd.read_csv(f'{data_folder}/before_finished/step_5/potential_actives.csv', sep=',', header=0)
potential_inactives = pd.read_csv(f'{data_folder}/before_finished/step_5/potential_inactives.csv', sep=',', header=0)

In [ ]:
inorganic_actives_cids, organic_actives_cids = filters.inorganic_filter(potential_actives, smi_col, cid_col, type='smiles')
print(f'Among actives, there are {len(organic_actives_cids)} organic molecules and {len(inorganic_actives_cids)} inorganic molecules')

inorganic_inactives_cids, organic_inactives_cids = filters.inorganic_filter(potential_inactives, smi_col, cid_col, type='smiles')
print(f'In inactives, there are {len(organic_inactives_cids)} organic molecules and {len(inorganic_inactives_cids)} inorganic molecules')

In [ ]:
#Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_7'):
    os.makedirs(f'{data_folder}/before_finished/step_7')

with open(f'{data_folder}/before_finished/step_7/inorganic.txt', 'w') as f:
    f.write("Actives:\n")
    for item in inorganic_actives_cids:
        f.write("%s\n" % item)
    f.write("\n\nInactives:\n")
    for item in inorganic_inactives_cids:
        f.write("%s\n" % item)

# Drop inorganics: 
potential_actives = potential_actives[~potential_actives[cid_col].isin(inorganic_actives_cids)]
potential_inactives = potential_inactives[~potential_inactives[cid_col].isin(inorganic_inactives_cids)]

#save: 
potential_actives.to_csv(f'{data_folder}/before_finished/step_7/organic_actives.csv', index=False)
potential_inactives.to_csv(f'{data_folder}/before_finished/step_7/organic_inactives.csv', index=False)

print('Dropped inorganic and saved organic compounds into step_7 folder.')

# 8. Mixture Handling

## 8.1. Quick Check

In [ ]:
#import: 
organic_actives = pd.read_csv(f'{data_folder}/before_finished/step_7/organic_actives.csv', sep=',', header=0)
organic_inactives = pd.read_csv(f'{data_folder}/before_finished/step_7/organic_inactives.csv', sep=',', header=0)

In [ ]:
filters.quick_check_mixtures('actives', organic_actives[smi_col])
filters.quick_check_mixtures('inactives', organic_inactives[smi_col])

Total number of mixtures in hits is 13
Total number of mixtures in decoys is 1906


## 8.2. Handling Mixtures

In [ ]:
processed_actives, removed_actives, small_organic_actives, small_inorganic_actives, not_lipinski_actives, cleaned_actives = filters.process_smi_mixtures(organic_actives, smi_col, cid_col)
processed_inactives, removed_inactives, small_organic_inactives, small_inorganic_inactives, not_lipinski_inactives, cleaned_inactives = filters.process_smi_mixtures(organic_inactives, smi_col, cid_col)

Cannot decide between ['CC1=NC2=C(O1)C3=CC=CC=C3C=C2', 'C1=C(C=C(C(=C1[N+](=O)[O-])O)[N+](=O)[O-])[N+](=O)[O-]'] for CID 3241713
Big organic molecule for CID 3244813 does not pass Lipinski's rule of five
Cannot decide between ['CN(C)C1=NC2=CC=CC=C2C(=C1)N', 'C1=C(NC(=O)NC1=O)C(=O)O'] for CID 646688
Cannot decide between ['C1C2(CN3CN1CN(C2)C3)N', 'C1=CC(=CN=C1)C(=O)O'] for CID 648270
Cannot decide between ['CN(C)C1=NC2=C(CCC2)C(=C1)N', 'C1=C(C=NC=C1O)C(=O)O'] for CID 652177
Cannot decide between ['CC1=NC2=C(S1)C3=CC=CC=C3C=C2', 'C1=C(C=C(C(=C1[N+](=O)[O-])O)[N+](=O)[O-])[N+](=O)[O-]'] for CID 654127
Cannot decide between ['C1CC1C(C2CC2)NC3=NCCO3', 'C(=C/C(=O)O)\\C(=O)O'] for CID 5388964
Cannot decide between ['CN1C(=O)C2=C(N=C(N2)Cl)N(C1=O)C(=O)[O-]', 'CN(C)CCOC(C1=CC=CC=C1)C2=CC=CC=C2'] for CID 657227
Cannot decide between ['CC[N+](C)(C)CC1=CC=CC=C1Br', 'CC1=CC=C(C=C1)S(=O)(=O)[O-]'] for CID 6100
Cannot decide between ['C1CN(CCN1CCOCCO)C(C2=CC=CC=C2)C3=CC=C(C=C3)Cl', 'C1=CC=C2C(=C1)C=C

Before updating the dataset, need to update the InChI for the mixtures.

In [ ]:
# Create a new step folder
if not os.path.exists(f'{data_folder}/before_finished/step_8'):
    os.makedirs(f'{data_folder}/before_finished/step_8')

#Generate df with the smiles column in the cleaned_actives or cleaned_inactives dictionary:
cleaned_actives_df = pd.DataFrame(list(cleaned_actives.values()), columns=[smi_col])
cleaned_inactives_df = pd.DataFrame(list(cleaned_inactives.values()), columns=[smi_col])

#Export the cleaned hits and inactives to csv:
cleaned_actives_df.to_csv(f'{data_folder}/before_finished/step_8/cleaned_mixtures_actives.csv', index=False, header=False)
cleaned_inactives_df.to_csv(f'{data_folder}/before_finished/step_8/cleaned_mixtures_inactives.csv', index=False, header=False)

In [ ]:
processed_hits_df = filters.process_mixture_df('actives_M1_agonist', organic_actives, processed_actives, removed_actives, small_organic_actives, small_inorganic_actives, smi_col, cid_col)
processed_inactives_df = filters.process_mixture_df('inactives_M1_agonist', organic_inactives, processed_inactives, removed_inactives, small_organic_inactives, small_inorganic_inactives, smi_col, cid_col)

In [ ]:
with open(f'{data_folder}/before_finished/step_8/mixture.txt', 'w') as f:
    f.write(f"""
Hits before processing: {len(organic_actives)}
Hits before processing: {len(organic_actives)}
Hits after processing: {len(processed_hits_df)}
Mixtures detected: {len(removed_actives)}
Mixtures with small inorganic molecules: {len(small_inorganic_actives)}
Mixtures with big organic molecules passing Lipinski: {len(small_organic_actives)}
Mixtures with big organic molecules not passing Lipinski: {len(not_lipinski_actives)}

Inactives before processing: {len(organic_inactives)}
Inactives after processing: {len(processed_inactives_df)}
Mixtures detected: {len(removed_inactives)}
Mixtures with small inorganic molecules: {len(small_inorganic_inactives)}
Mixtures with big organic molecules passing Lipinski: {len(small_organic_inactives)}
Mixtures with big organic molecules not passing Lipinski: {len(not_lipinski_inactives)}
""")

# Save the processed dataframes to csv
processed_hits_df.to_csv(f'{data_folder}/before_finished/step_8/post8_actives.csv', index=False)
processed_inactives_df.to_csv(f'{data_folder}/before_finished/step_8/post8_inactives.csv', index=False)

print('Dataframes saved successfully')

# 9. Neutralize & 10. Aromatize Molecules

In [ ]:
# Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_9_10'):
    os.makedirs(f'{data_folder}/before_finished/step_9_10')

#import:
pre9_actives = pd.read_csv(f'{data_folder}/before_finished/step_8/post8_actives.csv', sep=',', header=0)
pre9_inactives = pd.read_csv(f'{data_folder}/before_finished/step_8/post8_inactives.csv', sep=',', header=0)

In [ ]:
updated_smi = []

#Update dataset with neutralized, aromatic SMILES
for smi in pre9_actives[smi_col]: 
    mol = Chem.MolFromSmiles(smi)
    mol_neu = utils.neutralize_atoms(mol)
    smi_arom = utils.aromatize_smile(mol_neu)
    updated_smi.append(smi_arom)
    
#update the smiles in this df
pre9_actives[smi_col] = updated_smi

In [ ]:
updated_smi = []

#Update dataset with neutralized, aromatic SMILES
for smi in pre9_inactives[smi_col]: 
    mol = Chem.MolFromSmiles(smi)
    mol_neu = utils.neutralize_atoms(mol)
    smi_arom = utils.aromatize_smile(mol_neu)
    updated_smi.append(smi_arom)

#update the smiles in this df
pre9_inactives[smi_col] = updated_smi

In [ ]:
#Save
pre9_actives.to_csv(f'{data_folder}/before_finished/step_9_10/post10_actives.csv', index=False)
pre9_inactives.to_csv(f'{data_folder}/before_finished/step_9_10/post10_inactives.csv', index=False)

# Post 9+10: Update InChIs


Is it important to now update InChI in our datasets, for 2 reasons:

(1) Some mixture compounds have been modified (removal of small inorganic or organic molecules) in SMILES representation but not InChIs.

(2) The SMILES representations have been neutralized and aromatized, but not InChIs.

In [ ]:
#export the smiles columns to txt
pre9_actives[smi_col].to_csv(f'{data_folder}/before_finished/step_9_10/smiles_actives.txt', index=False, header=False)
pre9_inactives[smi_col].to_csv(f'{data_folder}/before_finished/step_9_10/smiles_inactives.txt', index=False, header=False)

In [ ]:
"""
Submit the smiles files to PubChem Identifier Exchange Service: 
    Input IDs: "SMILES"
    Operator type: "same CID" 
    Output IDs: "InChI"
    Output method: "Two column file showing each input output-correspondence"
    Compression: "No compression"
InChI list should be saved into "step_9_10" folder, named as "inchi_actives.txt" and "inchi_inactives" 
"""
#Import the converted InChIs
cleaned_inchi_hits = pd.read_csv(f'{data_folder}/before_finished/step_9_10/inchi_actives.txt', sep='\t', header=None)
cleaned_inchi_inactives = pd.read_csv(f'{data_folder}/before_finished/step_9_10/inchi_inactives.txt', sep='\t', header=None)

#a dictionary of smiles and corresponding inchi in cleaned_inchi_hits
hits_smi_inchi_dict = dict(zip(cleaned_inchi_hits[0], cleaned_inchi_hits[1]))
inactives_smi_inchi_dict = dict(zip(cleaned_inchi_inactives[0], cleaned_inchi_inactives[1]))
                             
#update the pre9_hits by matching the smiles with keys and replace inchi with values:
pre9_actives['InChI'] = pre9_actives[smi_col].map(hits_smi_inchi_dict) 
pre9_inactives['InChI'] = pre9_inactives[smi_col].map(inactives_smi_inchi_dict)

In [ ]:
#export: 
pre9_actives.to_csv(f'{data_folder}/before_finished/step_9_10/post10_actives.csv', index=False)
pre9_inactives.to_csv(f'{data_folder}/before_finished/step_9_10/post10_inactives.csv', index=False)

# 11. PAIN Filters

## 11.1. Frequency of Hits (FoH) Filter

Frequency of Hits is a complex concept that requires a merticulous approach. In general, the rule is if a compound was tested active in multiple assays, it is likely to be a promiscuous compound.
1. For each compounds, retrieve the information on its tested assays
2. For each of the assay tested, retrieve the sequence of the protein target.
3. Given all sequence of the protein tested, do a multiple sequence alignment to find the percentage Percent Identity (similarty) between these proteins. If an assay has high percentage to other targets, then these assays contribute less to promiscuousity of the compound.
4. Use the percentage identity as a weight:
w = 1 - %SI/100
Calculate the frequency of hits for each compound:
FoH = wACC/TAC
wACC is the weighed total number of assay tested where the compounds were identified acitives. TAC is the total number of assays tested.

In [ ]:
pre11_actives = pd.read_csv(f'{data_folder}/before_finished/step_9_10/post10_actives.csv', sep=',', header=0)
pre11_inactives = pd.read_csv(f'{data_folder}/before_finished/step_9_10/post10_inactives.csv', sep=',', header=0)

# Create a new step folder:
if not os.path.exists(f'{data_folder}/before_finished/step_11/11_1'):
    os.makedirs(f'{data_folder}/before_finished/step_11/11_1')

### 11.1.1. PubChem testing information for each compound

This part illustrates how to retrieve the information of how each compound was tested from the PubChem database. Bulk data retrieval from the ftp server is used to get the information of every bioassay in PubChem:

In [ ]:
url = 'https://ftp.ncbi.nlm.nih.gov/pubchem/Bioassay/Extras/bioassays.tsv.gz' #this FTP file records the summary data of all available AIDs in PubChem

local_save_dir = 'S:\coding\WelQrate\pubchem_sum'
local_save_path = os.path.join(local_save_dir, 'bioassays.tsv.gz')

if not os.path.exists(local_save_dir):
    os.makedirs(local_save_dir)
r = requests.get(url, stream=True)

with open(local_save_path, 'wb') as f:
    for chunk in r.iter_content(chunk_size=8192):
        f.write(chunk)
print('Downloaded to %s' % local_save_path)

path = 'pubchem_sum/bioassays.tsv.gz'

# Read the TSV file
all_bioassay = pd.read_csv(path, delimiter='\t')

In [ ]:
all_bioassay.head()

AID                                      BioAssay Name  Deposit Date  \
0    1  NCI human tumor cell line growth inhibition as...      20040815   
1    3  NCI human tumor cell line growth inhibition as...      20040815   
2    5  NCI human tumor cell line growth inhibition as...      20040815   
3    7  NCI human tumor cell line growth inhibition as...      20040815   
4    9  NCI human tumor cell line growth inhibition as...      20040815   

   Modify Date Source Name                                          Source ID  \
0     20231014     DTP/NCI  NCI human tumor cell line growth inhibition as...   
1     20231014     DTP/NCI  NCI human tumor cell line growth inhibition as...   
2     20231014     DTP/NCI  NCI human tumor cell line growth inhibition as...   
3     20231014     DTP/NCI  NCI human tumor cell line growth inhibition as...   
4     20231014     DTP/NCI  NCI human tumor cell line growth inhibition as...   

   Substance Type  Outcome Type Project Category   BioAssay Group  \
0  small-molecule  Confirmatory            Other  NCI-60_DOSERESP   
1  small-molecule  Confirmatory            Other  NCI-60_DOSERESP   
2  small-molecule  Confirmatory            Other  NCI-60_DOSERESP   
3  small-molecule  Confirmatory            Other  NCI-60_DOSERESP   
4  small-molecule  Confirmatory            Other  NCI-60_DOSERESP   

  BioAssay Types Protein Accessions UniProts IDs Gene IDs  Target TaxIDs  \
0            NaN                NaN          NaN      NaN            NaN   
1            NaN                NaN          NaN      NaN            NaN   
2            NaN                NaN          NaN      NaN            NaN   
3            NaN                NaN          NaN      NaN            NaN   
4            NaN                NaN          NaN      NaN            NaN   

  Taxonomy IDs  Number of Tested SIDs  Number of Active SIDs  \
0          NaN                  55004                   3280   
1          NaN                  51204                   2596   
2          NaN                  53832                   2486   
3          NaN                  53828                   4275   
4          NaN                  53744                   3125   

   Number of Tested CIDs  Number of Active CIDs  
0                  52994                   3057  
1                  49337                   2449  
2                  51804                   2301  
3                  51804                   4041  
4                  51771                   2948

### 11.1.2 Retrieving protein sequences for assays tested:

Then, the testing information for each compound is retrieved from the PugREST API

In [ ]:
# Cache to store the number of compounds tested per AID to avoid redundant call. 
num_compounds_tested_cache = {}

def get_num_compounds_tested(aid, all_bioassay=all_bioassay):
    """
    This function retrieves the information of how many compounds were tested in a given assay (by AID).
    """
    if aid in num_compounds_tested_cache:
        return num_compounds_tested_cache[aid]
    else: 
        #return the 'Number of Tested CIDs' column value at the row where the 'AID' column is equal to aid in the all_bioassay dataframe
        num_compounds_tested = all_bioassay[all_bioassay['AID'] == aid]['Number of Tested CIDs'].values[0]
    return num_compounds_tested

def get_assay_data(cid):
    """
    Return a dictionary of all targets that a given compound (by CID) was tested on in PubChem 
    and the activity values of the compound. 
    """
    url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/{cid}/assaysummary/JSON" #PUG-REST compound summary by CID
    response = requests.get(url)
    data = response.json()

    target_activity = {}

    if 'Table' in data and 'Row' in data['Table']:
        for row in data['Table']['Row']:
            cells = row['Cell']
            aid = int(cells[0])  # Extracting the AID from the first cell

            # Proceed only if the assay is a screening assay
            if cells[10] == 'Screening':

                # Proceed only if more than 10,000 compounds were tested
                num_compounds_tested = get_num_compounds_tested(aid)
                if num_compounds_tested > 10000:
                    target_gi = cells[5] # Retrieve the protein target's GI
                    activity_outcome = cells[4].lower()

                    if target_gi not in target_activity:
                        target_activity[target_gi] = activity_outcome == 'active'
                    elif activity_outcome == 'active':
                        target_activity[target_gi] = True # If a compound was tested multiple times on the same protein, priotize "active" outcome.
            
            else:
                continue

    return (cid, target_activity)

In [ ]:
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from tqdm import tqdm

cids_list = pre11_actives[cid_col].tolist()

def execute_with_multiprocessing(cids_list):
    """
    For a given list of CIDs, return a dictionary of dictionaries 
    of protein targets these compounds were tested on and the activity outcomes
    Input: 
        [list of CIDs]
    Output: 
        Dictionary of testing information for all CIDs, such as:
        {CID1:{target1:activity1, target3:activity3, ...},{CID2:{target2:activity2, target4:activity4, ...}, ...}}
    """
    results_dict = {}
    with ThreadPoolExecutor(max_workers=10) as executor:
        # Prepare futures for all CIDs
        futures = [executor.submit(get_assay_data, cid) for cid in cids_list]
        
        # Process futures as they complete
        for future in tqdm(as_completed(futures), total=len(cids_list), desc="Processing CIDs"):
            try:
                cid, target_activity = future.result()
                results_dict[cid] = target_activity
            except Exception as e:
                print(f"Error processing CID: {e}")
    return results_dict

results_dict = execute_with_multiprocessing(cids_list)

Processing CIDs:   0%|          | 0/188 [00:00<?, ?it/s]

Processing CIDs: 100%|██████████| 188/188 [01:20<00:00,  2.32it/s]


In [ ]:
#export results_dict
with open(f'{data_folder}/before_finished/step_11/11_1/results_dict.json', 'w') as f:
    json.dump(results_dict, f)

In [ ]:
#import results_dict:
with open(f'{data_folder}/before_finished/step_11/11_1/results_dict.json', 'r') as f:
    results_dict = json.load(f)

#get the list of all keys of the values in the dictionary:
protein_ids = []
for value in results_dict.values():
    protein_ids.extend(value.keys())

#clean the list
protein_ids = list(set(protein_ids))
protein_ids = [id for id in protein_ids if id != '']

print(protein_ids)
print(f'Require multiple sequencing alignment for {len(protein_ids)} proteins.')

['116292172', '78486550', '2507196', '2853980', '32425330', '115347926', '5454102', '148378801', '2393947', '6680530', '190938', '180352', '21595511', '15610402', '9629361', '597517618', '68989256', '4503385', '536029', '83627717', '9955963', '21392848', '12644416', '6831552', '118764400', '81899072', '119622516', '339641', '285809906', '735367775', '14149746', '1655766739', '20072248', '15610945', '31881630', '63102437', '341916350', '12803275', '116076351', '68474550', '1762973', '41872631', '56202836', '4506243', '4758484', '21264324', '119580345', '85666113', '76496497', '119607129', '7108463', '47123300', '1628587', '88702791', '4506055', '7706135', '493539358', '27894344', '67191027', '67463989', '134142337', '115430235', '208342286', '47678551', '60391226', '1654220559', '149631', '23110962', '1730321', '166202459', '86990435', '4826834', '4758878', '38027923', '4502331', '1927', '23893668', '998701', '27368096', '23943882', '110611243', '21595776', '28373962', '37589898', '5641

Now we retrieve all the FASTA sequences of proteins tested for all of our compounds with Biopython API to Entrez of NCBI. The FASTA sequence is saved as "sequences.fasta"

In [ ]:
# Always tell NCBI who you are
Entrez.email = "hdong26@amherst.edu"

# The filename where you want to save the sequences
output_filename = f'{data_folder}/before_finished/step_11/11_1/sequences.fasta'

# Open a file to write the sequences
with open(output_filename, "w") as output_file:
    for id in protein_ids:
        try:
            # Fetch the sequence from NCBI
            handle = Entrez.efetch(db="protein", id=id, rettype="fasta", retmode="text")
            sequence_data = handle.read()
            handle.close()
            
            # Write the sequence data to the file
            output_file.write(sequence_data)
        except Exception as e:
            print(f"An error occurred while fetching {id}: {e}")

From the FASTA sequence, we also need to retrieve the list of protein names, since these are different from the protein GIs.

In [ ]:
def extract_protein_names(file_path):
    protein_names = []
    with open(file_path, 'r') as file:
        for line in file:
            if line.startswith('>'):
                # Split the line at spaces and take the first item
                parts = line.split(' ')
                protein_name = parts[0]
                # Remove the leading '>' character
                protein_name = protein_name[1:]
                protein_names.append(protein_name)
    return protein_names

file_path = f'{data_folder}/before_finished/step_11/11_1/sequences.fasta'
protein_names = extract_protein_names(file_path)

# Create a dictionary to map protein IDs to protein names by index 
protein_id_to_name = {protein_ids[i]: protein_names[i] for i in range(len(protein_ids))}

In [ ]:
print(protein_id_to_name)

{'116292172': 'NP_001013728.2', '78486550': 'NP_598230.2', '2507196': 'sp|P53779.2|MK10_HUMAN', '2853980': 'AAC83551.1', '32425330': 'AAH19268.2', '115347926': 'CAL20848.1', '5454102': 'NP_006333.1', '148378801': 'YP_001253342.1', '2393947': 'AAC51766.1', '6680530': 'NP_032451.1', '190938': 'AAA36557.1', '180352': 'AAA51985.1', '21595511': 'AAH32261.1', '15610402': 'NP_217783.1', '9629361': 'NP_057851.1', '597517618': 'CDO16711.1', '68989256': 'NP_004283.2', '4503385': 'NP_000786.1', '536029': 'CAA84846.1', '83627717': 'NP_031943.3', '9955963': 'NP_005680.1', '21392848': 'NP_652928.1', '12644416': 'sp|Q14749.3|GNMT_HUMAN', '6831552': 'sp|O75388.1|GPR32_HUMAN', '118764400': 'AAI28575.1', '81899072': 'sp|Q8C6L5.1|CGAS_MOUSE', '119622516': 'EAX02111.1', '339641': 'AAC63054.1', '285809906': 'DAA06693.1', '735367775': 'NP_000202.3', '14149746': 'NP_066285.1', '1655766739': 'NP_001357589.1', '20072248': 'AAH26347.1', '15610945': 'NP_218326.1', '31881630': 'NP_000947.2', '63102437': 'AAH95408

### 11.1.3. Percent Sequence Identity by Multiple Sequence Alignment

The sequences.fasta file is submitted to https://www.ebi.ac.uk/jdispatcher/msa/clustalo for multiple sequencing alignment. The resulted table of percent sequence identity matrix is saved and imported for the calculation of FoH

In [ ]:
"""
Submit sequences.fasta to https://www.ebi.ac.uk/jdispatcher/msa/clustalo
    Input sequence type: Protein
    Output format: ClustalW with character counts
Download the resulted Percent Identity Matrix file file and save as "percent_identity_matrix.txt"
"""

#import the identity matrix:
protein_si = pd.read_csv(
    f'{data_folder}/before_finished/step_11/11_1/percent_identity_matrix.txt',
    delimiter='\s+',
    header=None,
    skiprows=6 
)

In [ ]:
#remove the first column:
protein_si = protein_si.drop(protein_si.columns[0], axis=1)

name = protein_si[1].tolist()
name = ['protein name'] + name
protein_si.columns = name

In [ ]:
protein_si

protein name  AAH65243.1  NP_001341533.1  NP_937802.1  AAA34498.1  \
0        AAH65243.1      100.00           97.37        97.34       13.33   
1    NP_001341533.1       97.37          100.00       100.00       13.33   
2       NP_937802.1       97.34          100.00       100.00       13.33   
3        AAA34498.1       13.33           13.33        13.33      100.00   
4        pdb|1ULL|B       23.08           23.08        23.08       17.65   
..              ...         ...             ...          ...         ...   
420     NP_066268.1        7.41            7.87         8.14        0.00   
421     NP_057231.1        8.79           10.99        10.99        6.25   
422  YP_001253342.1       13.01           12.31        12.60       11.11   
423     XP_718648.1        2.50            8.47         8.62         NaN   
424      AAF25870.1        3.12            7.38         7.44        0.00   

     pdb|1ULL|B  AAA36557.1  AAB23646.1  NP_000469.3  NP_001074551.1  ...  \
0         23.08       10.53       33.33        12.38           14.29  ...   
1         23.08       10.53       33.33        10.64           14.97  ...   
2         23.08       10.53       33.33        10.64           14.97  ...   
3         17.65       21.05       25.00        11.11           11.11  ...   
4        100.00       17.65        0.00         0.00            0.00  ...   
..          ...         ...         ...          ...             ...  ...   
420        0.00        8.33       11.11         9.52            9.64  ...   
421       10.00       10.00        9.09        13.00           13.51  ...   
422        8.33       14.81        7.14        10.85           12.33  ...   
423         NaN        0.00         NaN        15.07           13.70  ...   
424        0.00       23.53       14.29         8.13           11.57  ...   

     NP_000762.2  NP_001970.2  NP_058021.1  XP_001348533.1  \
0          14.63        10.14        16.36            6.45   
1          13.56         9.17        17.91            4.96   
2          13.91         9.40        17.91            5.08   
3            NaN         0.00        10.00            0.00   
4            NaN         0.00        11.11            0.00   
..           ...          ...          ...             ...   
420        13.40        14.49        11.98           23.05   
421        13.43        16.13        13.98            7.04   
422        19.29        14.73        20.22           17.55   
423        11.72        10.62        12.02           13.85   
424         9.75        11.08        16.79           14.51   

     sp|P08753.3|GNAI3_RAT  NP_066268.1  NP_057231.1  YP_001253342.1  \
0                     5.66         7.41         8.79           13.01   
1                     7.95         7.87        10.99           12.31   
2                     8.24         8.14        10.99           12.60   
3                     0.00         0.00         6.25           11.11   
4                     0.00         0.00        10.00            8.33   
..                     ...          ...          ...             ...   
420                  71.39       100.00        12.00           16.97   
421                  18.00        12.00       100.00           19.45   
422                  18.05        16.97        19.45          100.00   
423                  10.95        10.43        16.38           19.00   
424                  10.43         9.35        10.75           17.72   

     XP_718648.1  AAF25870.1  
0           2.50        3.12  
1           8.47        7.38  
2           8.62        7.44  
3            NaN        0.00  
4            NaN        0.00  
..           ...         ...  
420        10.43        9.35  
421        16.38       10.75  
422        19.00       17.72  
423       100.00       19.34  
424        19.34      100.00  

[425 rows x 426 columns]

### 11.1.4. Calculation of FoHs:

Until now, we have a dictionary of (cid: assays tested); (assay_tested:protein name), and percentage identity matrix with first columns as protein names.
For each compound, we retrieve the list of all protein names tested on that compounds by matching between the two first dictionary. From this list, we retrieve the corresponding matrix of percentages identitiy of these proteins corresponding to these compounds and calculate the FoH

In [ ]:
protein_si_dict = {}
for name in protein_si['protein name']: 
    for other_name in protein_si['protein name']: 
        if other_name != name: 
            protein_si_dict[(name, other_name)] = protein_si.loc[protein_si['protein name'] == name, other_name].values[0]

In [ ]:
foh_dict = {}

for cid, targets in tqdm.tqdm(results_dict.items()):
    active_weight_list = []
    total_weight_list = []

    for target_id, result in targets.items():
        if target_id == '':
            continue

        protein_name = protein_id_to_name[target_id]
        max_weight = 0

        for other_id, other_result in targets.items():
            if other_id != target_id and other_id != '':
                other_protein_name = protein_id_to_name[other_id]
                value = protein_si_dict[(protein_name, other_protein_name)]
                max_weight = max(max_weight, value)

        target_weight = 1 - max_weight / 100

        if result:
            active_weight_list.append(target_weight)
        total_weight_list.append(target_weight)

    if total_weight_list:
        foh_score = sum(active_weight_list) / sum(total_weight_list)
        foh_dict[cid] = foh_score
    else: 
        foh_dict[cid] = 0

100%|██████████| 188/188 [00:22<00:00,  8.39it/s]


In [ ]:
#export foh_dict
with open(f'{data_folder}/before_finished/step_11/11_1/foh_dict.json', 'w') as f:
    json.dump(foh_dict, f)

For compounds with FoH larger than 0.26, we remove them

In [ ]:
to_drop = []
for cid, foh_score in foh_dict.items():
    if foh_score > 0.26: 
        to_drop.append(cid)

post_FoH_actives = pre11_actives[~pre11_actives[cid_col].isin(to_drop)]
print(f'Dropped {(len(to_drop))} compounds with FoH larger than 0.26')

Dropped 1 compounds with FoH larger than 0.26


In [ ]:
post_FoH_actives.to_csv(f'{data_folder}/before_finished/step_11/11_1/post_FoH_actives.csv', index=False)
pre11_inactives.to_csv(f'{data_folder}/before_finished/step_11/11_1/post_FoH_inactives.csv', index=False)

## 11.2. Autofluoresence Filter

In [ ]:
post_FoH_actives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_1/post_FoH_actives.csv', sep=',', header=0)
post_FoH_inactives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_1/post_FoH_inactives.csv', sep=',', header=0)

When finding false positive due to autofluorescence and luceferase inhibition, it is important to check if the particular assays use one of these technologies. Here, all three assays (AID626, AID1488, and AID1741) use fluorescence technologies, so it is optimal to remove compounds that are active in AIDs: 587, 588, 590, 591, 592, 593, 594

In [3]:
autofluorescence_cids = filters.load_autofluorescence_cids(data_folder)

Autofluorescence compounds imported successfully!


In [ ]:
to_drop_hits = []
for cid in post_FoH_actives:
    if cid in autofluorescence_cids:
        to_drop.append(cid)
post_autofluorescence_actives = post_FoH_actives[~post_FoH_actives['PUBCHEM_CID'].isin(to_drop_hits)]
print(f'Dropped {(len(to_drop_hits))} autofluorescence compounds')

Dropped 0 autofluorescence compounds


In [ ]:
post_autofluorescence_actives

PUBCHEM_CID                      PUBCHEM_EXT_DATASOURCE_SMILES  \
0        2882111                  Cc1ccc(C)c(OCC(O)CN2C(C)CCCC2C)c1   
1        2328849              CCOC(=O)N1CCN(C2=Nc3cccc4cccc2c34)CC1   
2        2321890  CC(=O)Nc1ccc(S(=O)(=O)Nc2nc3ccccc3nc2Sc2nncn2C...   
3        2310918  C=CCn1c(=O)/c(=C/c2ccc(O)cc2)s/c1=C(/C#N)c1nnc...   
4        2548564  CCN(CC(=O)Nc1ccc2c(c1)OCCO2)c1nc(-c2cccnc2)nc2...   
..           ...                                                ...   
183      6602803                  CCc1nc(NCc2ccccn2)c2oc3ccccc3c2n1   
184       664440  CCOC(=O)c1cc2c(=O)n3cccc(C)c3nc2n(Cc2ccco2)c1=...   
185       664704  Cc1cccn2c(=O)c3cc(C#N)c(=NC(=O)c4cccnc4)n(CC4C...   
186       664895                CC(C)OCCCNc1ncnc2c1[nH]c1ccc(F)cc12   
187       665546  COC(=O)c1c(NC(=O)CN2CCN(c3ccccc3)CC2)c2c(C)ccc...   

    PUBCHEM_ACTIVITY_OUTCOME  \
0                     Active   
1                     Active   
2                     Active   
3                     Active   
4                     Active   
..                       ...   
183                   Active   
184                   Active   
185                   Active   
186                   Active   
187                   Active   

                                                 InChI  \
0    InChI=1S/C18H29NO2/c1-13-8-9-14(2)18(10-13)21-...   
1    InChI=1S/C18H19N3O2/c1-2-23-18(22)21-11-9-20(1...   
2    InChI=1S/C19H17N7O3S2/c1-12(27)21-13-7-9-14(10...   
3    InChI=1S/C22H21N5O2S/c1-2-11-27-21(29)18(13-15...   
4    InChI=1S/C25H23N5O3/c1-2-30(16-23(31)27-18-9-1...   
..                                                 ...   
183  InChI=1S/C18H16N4O/c1-2-15-21-16-13-8-3-4-9-14...   
184  InChI=1S/C26H21N5O5/c1-3-35-26(34)20-13-19-22(...   
185  InChI=1S/C24H20N6O3/c1-15-5-3-9-29-20(15)27-22...   
186  InChI=1S/C16H19FN4O/c1-10(2)22-7-3-6-18-16-15-...   
187  InChI=1S/C24H28N4O3/c1-17-8-7-11-19-21(17)22(2...   

    Mol removed from mixture Small inorganic molecule  
0                         Cl                       Cl  
1                        NaN                      NaN  
2                        NaN                      NaN  
3                        NaN                      NaN  
4                        NaN                      NaN  
..                       ...                      ...  
183                       Cl                       Cl  
184                      NaN                      NaN  
185                      NaN                      NaN  
186                      NaN                      NaN  
187                      NaN                      NaN  

[187 rows x 6 columns]

In [ ]:
if not os.path.exists(f'{data_folder}/before_finished/step_11/11_2'):
    os.makedirs(f'{data_folder}/before_finished/step_11/11_2')

In [ ]:
#save:
post_autofluorescence_actives.to_csv(f'{data_folder}/before_finished/step_11/11_2/post_autofluorescence_actives.csv', index=False)
post_FoH_inactives.to_csv(f'{data_folder}/before_finished/step_11/11_2/post_autofluorescence_inactives.csv', index=False)

## 11.3 RDKit PAIN Filter

In [ ]:
post_autofluorescence_actives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_2/post_autofluorescence_actives.csv', sep=',', header=0)
post_autofluorescence_inactives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_2/post_autofluorescence_inactives.csv', sep=',', header=0)

In [ ]:
pains_actives = filters.detect_pains(post_autofluorescence_actives, smi_col, cid_col)

1 pains detected
2 pains detected
3 pains detected
4 pains detected
5 pains detected
6 pains detected
7 pains detected
8 pains detected
9 pains detected
10 pains detected
11 pains detected
12 pains detected
13 pains detected
14 pains detected
15 pains detected
16 pains detected
17 pains detected
18 pains detected


In [ ]:
post_pains_actives = post_autofluorescence_actives[~post_autofluorescence_actives['PUBCHEM_CID'].isin(pains_actives)]

In [ ]:
if not os.path.exists(f'{data_folder}/before_finished/step_11/11_3'):
    os.makedirs(f'{data_folder}/before_finished/step_11/11_3')

post_pains_actives.to_csv(f'{data_folder}/before_finished/step_11/11_3/post_pains_actives.csv', index=False)
post_autofluorescence_inactives.to_csv(f'{data_folder}/before_finished/step_11/11_3/post_pains_inactives.csv', index=False)

# 12. Drug-likeness Filter:

In [ ]:
pre12_actives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_3/post_pains_actives.csv', sep=',', header=0)
pre12_inactives = pd.read_csv(f'{data_folder}/before_finished/step_11/11_3/post_pains_inactives.csv', sep=',', header=0)

In [ ]:
not_drug_actives = filters.drug_likeness_filter_multiprocessing(pre12_actives, smi_col, cid_col)

Processing SMILES: 100%|██████████| 169/169 [00:00<?, ?it/s]


In [ ]:
not_drug_inactives = filters.drug_likeness_filter_multiprocessing(pre12_inactives, smi_col, cid_col)

Processing SMILES: 100%|██████████| 61725/61725 [00:10<00:00, 5705.98it/s]  


In [ ]:
post12_actives = pre12_actives[~pre12_actives['PUBCHEM_CID'].isin(not_drug_actives)]
post12_inactives = pre12_inactives[~pre12_inactives['PUBCHEM_CID'].isin(not_drug_inactives)]

print(f'Dropped {(len(not_drug_actives))} hit compounds that do not pass the drug likeness filter')
print(f'Dropped {(len(not_drug_inactives))} inactive compounds that do not pass the drug likeness filter')

Dropped 4 hit compounds that do not pass the drug likeness filter
Dropped 1169 decoy compounds that do not pass the drug likeness filter


In [ ]:
if not os.path.exists(f'{data_folder}/before_finished/step_12'):
    os.makedirs(f'{data_folder}/before_finished/step_12')

#Export not_drug_actives and inactives:
with open(f'{data_folder}/before_finished/step_12/not_drug_actives.json', 'w') as f:
    json.dump(not_drug_actives, f)
with open(f'{data_folder}/before_finished/step_12/not_drug_inactives.json', 'w') as f:
    json.dump(not_drug_inactives, f)

In [ ]:
# save: 
post12_actives.to_csv(f'{data_folder}/before_finished/step_12/post12_actives.csv', index=False)
post12_inactives.to_csv(f'{data_folder}/before_finished/step_12/post12_inactives.csv', index=False)

# 13. ChemBL Curation Pipeline

Besides PubChem, the ChEMBL database is one of several public databases containing bioactivity data on small molecule compounds curated from various sources. Incoming compounds are typically not standardized according to consistent rules. To maintain the quality of the final database and to facilitate the comparison and integration of data on the same compound from different sources, it is essential to appropriately standardize the chemical structures in the database. This chemical curation pipeline has been developed by Bento, A.P, et al., using the open-source toolkit RDKit, including a Checker module that tests the validity of chemical structures and flags any serious errors. For ChEMBL, a penalty score of 7 is considered a fatal error, and the molfile is not loaded into the database.

In our dataset, we passed all compounds through this Checker module to evaluate the validity of the included chemical structures. For compounds that returned a penalty score of 7, we will check them manually.

Reference:
Bento, A.P., Hersey, A., Félix, E. et al. An open source chemical structure curation pipeline using RDKit. J Cheminform 12, 51 (2020). https://doi.org/10.1186/s13321-020-00456-1

In [ ]:
pre13_actives = pd.read_csv(f'{data_folder}/before_finished/step_12/post12_actives.csv', sep=',', header=0)
pre13_inactives = pd.read_csv(f'{data_folder}/before_finished/step_12/post12_inactives.csv', sep=',', header=0)

In [ ]:
# Apply the ChemBL Curation Pipeline Checker module:
score_actives = utils.checker_multiprocessing(pre13_actives, smi_col, cid_col)
score_inactives = utils.checker_multiprocessing(pre13_inactives, smi_col, cid_col)

Processing SMILES: 100%|██████████| 60556/60556 [00:42<00:00, 1412.51it/s] 


In [ ]:
#print all unique values in the dictionary:
print(set(score_actives.values()))
print(set(score_inactives.values()))

{0, 2}
{0, 2, 5, 6}


In [ ]:
# Create a new step folder
if not os.path.exists(f'{data_folder}/before_finished/step_13'):
    os.makedirs(f'{data_folder}/before_finished/step_13')

# save the scores:
with open(f'{data_folder}/before_finished/step_13/score_actives.json', 'w') as f:
    json.dump(score_actives, f)
with open(f'{data_folder}/before_finished/step_13/score_inactives.json', 'w') as f:
    json.dump(score_inactives, f)

In [ ]:
#drop all compounds with a penalty score of 7:
to_drop_actives = []
to_drop_inactives = []
for cid, penalty_score in score_actives.items():
    if penalty_score == 7:
        to_drop_actives.append(cid)
for cid, penalty_score in score_inactives.items():
    if penalty_score == 7:
        to_drop_inactives.append(cid)

post13_actives = pre13_actives[~pre13_actives[cid_col].isin(to_drop_actives)]
post13_inactives = pre13_inactives[~pre13_inactives[cid_col].isin(to_drop_inactives)]

In [ ]:
#save final_hits and inactives:
post13_actives.to_csv(f'{data_folder}/before_finished/step_13/post13_actives.csv', index=False)
post13_inactives.to_csv(f'{data_folder}/before_finished/step_13/post13_inactives.csv', index=False)

# 14. Final handling of chemical representations

Any kind of molecular processing should come with a special attention to post-processing adjustment. This is because during molecular processing, changes in data representation and format could lead to inconsistency or redundancy in the datasets. Below are the two adjustments performed on our datasets. 

(1) Some of the InChIs in our datasets should be updated. This is because some of the InChI will be missing, since the PubChem Identifier Exchange service might not able to find the corresponding InChI for the aromatized, neutralized SMILES.

(2) Presence of some additional duplicates resulted from molecular processing: While handling mixtures, there might be some mixtures whose component molecules are identical. For example, mixtures of organic and inorganic molecules (molX-ionA and molX-ionB) after removed the ions (ionA and ionB) will result in duplicates (molX). Moreover, since the original mixtures are different, their activities could be different. Therefore, we also need to check their activities while handling these duplicates.
- If all duplicates share the same results (active/inactive), we keep one of them, since it is likely that the organic molecule kept contributed more significantly to the activity of the mixture. 
- If duplicates of the same molecules returned different activity, it is safer to remove both of them.

In [ ]:
#import: 
pre14_actives = pd.read_csv(f'{data_folder}/before_finished/step_13/post13_actives.csv', sep=',', header=0)
pre14_inactives = pd.read_csv(f'{data_folder}/before_finished/step_13/post13_inactives.csv', sep=',', header=0)

## 14.1 Update InChI

In [ ]:
def smi_to_inchi(smi):
    mol = Chem.MolFromSmiles(smi)
    inchi = Chem.inchi.MolToInchi(mol)
    return inchi

In [ ]:
count = 0 
for index, row in pre14_actives.iterrows():
    if row['InChI'] != row['InChI']:
        pre14_actives.at[index, 'InChI'] = smi_to_inchi(row[smi_col])
        count += 1
print(f'Updated {count} InChI values in pre14_actives')

count = 0
for index, row in pre14_inactives.iterrows():
    if row['InChI'] != row['InChI']:
        pre14_inactives.at[index, 'InChI'] = smi_to_inchi(row[smi_col])
        count += 1
print(f'Updated {count} InChI values in pre14_inactives')

Updated 0 InChI values in pre13_hits


[14:56:23] WARNING: Charges were rearranged

[14:56:24] WARNING: Omitted undefined stereo

[14:56:24] WARNING: Omitted undefined stereo

[14:56:24] WARNING: Omitted undefined stereo

[14:56:24] WARNING: Omitted undefined stereo

[14:56:25] WARNING: Omitted undefined stereo

[14:56:25] WARNING: Omitted undefined stereo



Updated 25 InChI values in pre13_decoys


[14:56:25] WARNING: Omitted undefined stereo



## 14.2 Handle duplicates

In [ ]:
#Check if a mol in active set appeared in inactive set:
for i in pre14_actives[smi_col]:
    if i in list(pre14_inactives[smi_col]):
        print(f'{i} SMILES appeared in both active and inactive sets')
for i in pre14_actives['InChI']:
    if i in list(pre14_inactives['InChI']):
        print(f'{i} InChI appeared in both active and inactive sets')

#Return all duplicates by comparing InChI:
final_actives_duplicates_InChI = pre14_actives[pre14_actives.duplicated(subset=['InChI'], keep=False)]
final_inactives_duplicates_InChI = pre14_inactives[pre14_inactives.duplicated(subset=['InChI'], keep=False)]
final_actives_duplicates_smi = pre14_actives[pre14_actives.duplicated(subset=[smi_col], keep=False)]
final_inactives_duplicates_smi = pre14_inactives[pre14_inactives.duplicated(subset=[smi_col], keep=False)]

print('Number of InChI duplicates in actives: ', len(final_actives_duplicates_InChI))
print('Number of InChI duplicates in inactives: ', len(final_inactives_duplicates_InChI))
print('Number of SMILES duplicates in actives: ', len(final_actives_duplicates_smi))
print('Number of SMILES duplicates in inactives: ', len(final_inactives_duplicates_smi))

In [ ]:
if not os.path.exists(f'{data_folder}/before_finished/step_14'):
    os.makedirs(f'{data_folder}/before_finished/step_14')

#write all the duplicates to a file:
#write duplicates to a txt file: 
with open(f'{data_folder}/before_finished/step_14/duplicates.txt', 'w') as f:
    f.write('InChI duplicates in actives: \n')
    f.write(final_actives_duplicates_InChI.to_string())
    f.write('\n\n')
    f.write('InChI duplicates in inactives: \n')
    f.write(final_inactives_duplicates_InChI.to_string())
    f.write('\n\n')
    f.write('SMILES duplicates in actives: \n')
    f.write(final_actives_duplicates_smi.to_string())
    f.write('\n\n')
    f.write('SMILES duplicates in inactives: \n')
    f.write(final_inactives_duplicates_smi.to_string())

In [ ]:
#remove these duplicates, keep the first one: 
#by inchi:
final_actives = pre14_actives.drop_duplicates(subset=['InChI'], keep='first')
final_inactives = pre14_inactives.drop_duplicates(subset=['InChI'], keep='first')

In [ ]:
if len(final_actives[final_actives.duplicated(subset=[smi_col], keep=False)]) == 0:
    print('No more duplicates in actives')

if len(final_inactives[final_inactives.duplicated(subset=[smi_col], keep=False)]) == 0:
    print('No more duplicates in inactives')

In [ ]:
# save: 
if not os.path.exists(f'{data_folder}/finished'):
    os.makedirs(f'{data_folder}/finished')
final_actives.to_csv(f'{data_folder}/finished/final_actives.csv',sep=',', index=False)
final_inactives.to_csv(f'{data_folder}/finished/final_inactives.csv',sep=',', index=False)

# Some Additional Modifications

## A. Adjust Column Names

In [ ]:
final_hits = pd.read_csv(f'{data_folder}/finished/final_hits.csv', sep=',', header=0)
final_inactives = pd.read_csv(f'{data_folder}/finished/final_inactives.csv', sep=',', header=0)

In [ ]:
# Add another column: "activity_value" with all empty NaN values for format consistency with regression datasets:
final_hits.loc[:, 'activity_value'] = np.nan
final_inactives.loc[:, 'activity_value'] = np.nan

if 'small_organic_mol_from_mixture' not in final_actives.columns: # Sometimes, this column might be missing since no small org. mol was detected in mixtures.
    final_actives['small_organic_mol_from_mixture'] = np.nan

# Rename the columns:
final_hits = final_hits.rename(columns={
    'PUBCHEM_CID': 'CID',
    'PUBCHEM_ACTIVITY_OUTCOME': 'activity_outcome',
    'PUBCHEM_EXT_DATASOURCE_SMILES': 'SMILES',
    'Mol removed from mixture': 'mol_removed_from_mixture',
    'Small inorganic molecule': 'small_inorganic_mol_from_mixture',
    'Small organic molecule': 'small_organic_mol_from_mixture'
})
final_inactives = final_inactives.rename(columns={
    'PUBCHEM_CID': 'CID',
    'PUBCHEM_ACTIVITY_OUTCOME': 'activity_outcome',
    'PUBCHEM_EXT_DATASOURCE_SMILES': 'SMILES',
    'Mol removed from mixture': 'mol_removed_from_mixture',
    'Small inorganic molecule': 'small_inorganic_mol_from_mixture',
    'Small organic molecule': 'small_organic_mol_from_mixture'
})

In [ ]:
# If there is no "small_organic_mol_from_mixture" column, add it:
if 'small_organic_mol_from_mixture' not in final_hits.columns:
    final_hits.loc[:, 'small_organic_mol_from_mixture'] = np.nan
if 'small_organic_mol_from_mixture' not in final_inactives.columns:
    final_inactives.loc[:, 'small_organic_mol_from_mixture'] = np.nan

In [ ]:
#swap the positions of the columns InChI and activity_outcome:
final_hits = final_hits[['CID', 'SMILES', 'InChI', 'activity_outcome', 'activity_value', 'mol_removed_from_mixture', 'small_inorganic_mol_from_mixture', 'small_organic_mol_from_mixture']]
final_inactives = final_inactives[['CID', 'SMILES', 'InChI', 'activity_outcome', 'activity_value', 'mol_removed_from_mixture', 'small_inorganic_mol_from_mixture', 'small_organic_mol_from_mixture']]

In [ ]:
#export:
final_hits.to_csv(f'{data_folder}/finished/final_hits.csv', sep=',', index=False)
final_inactives.to_csv(f'{data_folder}/finished/final_inactives.csv', sep=',', index=False)

## B. Compile Control Data

This part of the notebook includes our complilation of the control data (poorly curated data without applying hierarchical curation, molecular processing, or filters). Data format should be similar to curated data for convenience.

In [ ]:
raw_M1_agonist = pd.read_csv(f'{data_folder}/before_finished/step_1/AID626.csv', sep=',', header=0)

In [ ]:
std_inchi = pd.read_csv(f'{data_folder}/before_finished/step_3/std_inchi_626.txt', sep='\t', header=None)

In [ ]:
#Convert AID626_InChI to a dictionary:
raw_M1_agonist_dict = dict(zip(std_inchi[0], std_inchi[1]))
#Add InChI to AID626 dataframe:
raw_M1_agonist['InChI'] = raw_M1_agonist['PUBCHEM_CID'].map(raw_M1_agonist_dict)

In [ ]:
raw_hits = raw_M1_agonist[raw_M1_agonist['PUBCHEM_ACTIVITY_OUTCOME'] == 'Active']
raw_inactives = raw_M1_agonist[raw_M1_agonist['PUBCHEM_ACTIVITY_OUTCOME'] == 'Inactive']

In [ ]:
# Add some other columns to match the format of the curated data:
raw_hits.loc[:, 'activity_value'] = np.nan
raw_hits.loc[:, 'mol_removed_from_mixture'] = np.nan
raw_hits.loc[:, 'small_inorganic_mol_from_mixture'] = np.nan
raw_hits.loc[:, 'small_organic_mol_from_mixture'] = np.nan

raw_inactives.loc[:, 'activity_value'] = np.nan
raw_inactives.loc[:, 'mol_removed_from_mixture'] = np.nan
raw_inactives.loc[:, 'small_inorganic_mol_from_mixture'] = np.nan
raw_inactives.loc[:, 'small_organic_mol_from_mixture'] = np.nan

In [ ]:
# Rename the columns:
raw_hits = raw_hits.rename(columns={
    'PUBCHEM_CID': 'CID',
    'PUBCHEM_ACTIVITY_OUTCOME': 'activity_outcome',
    'PUBCHEM_EXT_DATASOURCE_SMILES': 'SMILES',
})
raw_inactives = raw_inactives.rename(columns={
    'PUBCHEM_CID': 'CID',
    'PUBCHEM_ACTIVITY_OUTCOME': 'activity_outcome',
    'PUBCHEM_EXT_DATASOURCE_SMILES': 'SMILES',
})

In [ ]:
#swap the positions of the columns InChI and activity_outcome:
raw_hits = raw_hits[['CID', 'SMILES', 'InChI', 'activity_outcome', 'activity_value', 'mol_removed_from_mixture', 'small_inorganic_mol_from_mixture', 'small_organic_mol_from_mixture']]
raw_inactives = raw_inactives[['CID', 'SMILES', 'InChI', 'activity_outcome', 'activity_value', 'mol_removed_from_mixture', 'small_inorganic_mol_from_mixture', 'small_organic_mol_from_mixture']]

In [ ]:
if not os.path.exists(f'{data_folder}/finished/control_data'):
    os.makedirs(f'{data_folder}/finished/control_data')

#save the hits and inactives
raw_hits.to_csv(f'{data_folder}/finished/control_data/raw_hits.csv', sep=',', index=False)
raw_inactives.to_csv(f'{data_folder}/finished/control_data/raw_inactives.csv', sep=',', index=False)

#save as txt:
raw_hits.to_csv(f'{data_folder}/finished/control_data/raw_hits.txt', sep=';', index=False, header=False)
raw_inactives.to_csv(f'{data_folder}/finished/control_data/raw_inactives.txt', sep=';', index=False, header=False)